### PR TITLE
Search element by text

### PR DESCRIPTION
**Pull request by nkh**

### CODE REVIEW

File: lib/App/Asciio/stripes/group.pm [hunk 1]
```diff
@@ -38,8 +38,7 @@ for my $element (@{$elements})
 		my $text = $stripe->{TEXT} ;
 		
 		my $width = 0 ;
-		my $usc_length = usc_length($_) ;
-		map {$width  = $width < $usc_length ? $usc_length  : $width} split("\n", $text) ;
+		map {$width  = $width < usc_length($_) ? usc_length($_)  : $width} split("\n", $text) ;
```
Comment (nkh) [inline lib/App/Asciio/stripes/group.pm:41]: Good catch but you're computing the usc length twice and that an expensive operation.

blocks for map/grep, ... are normal blocks so you can declare variables in them, in this case to compute the usc_length just once.

File: lib/App/Asciio/stripes/stripes.pm [hunk 2]
```diff
@@ -22,9 +22,8 @@ for my $stripe (@{$element_definition->{STRIPES}})
 	my $text = $stripe->{TEXT} ;
 	
 	my $width = 0 ;
-	my $usc_length = usc_length($_) ;
 	
-	map { $width  = $width < $usc_length ? $usc_length : $width } split("\n", $text) ;
+	map { $width  = $width < usc_length($_) ? usc_length($_) : $width } split("\n", $text) ;
```
Comment (nkh) [inline lib/App/Asciio/stripes/stripes.pm:26]: same as above comment

File: lib/App/Asciio/Actions/ElementsManipulation.pm [hunk 3]
```diff
@@ -175,6 +175,18 @@ $self->select_all_elements() ;
 $self->update_display() ;
 }
 
+#----------------------------------------------------------------------------------------------
+
```
Comment (nkh) [inline lib/App/Asciio/Actions/ElementsManipulation.pm:179]: some empty lines make code easier to read

**Review by nkh — COMMENTED** [review]
reviewing code in git hub is horrible, lol! I get to use 1/3 of my screen

well, it's "practical" 

Good luck with the refactoring.  keep your old commits so you can make a diff after so you see what you've done better. I think this is a good way to learn some Perl.

File: lib/App/Asciio/stripes/group.pm (hunk 1 above)
Comment (qindapao) [reply lib/App/Asciio/stripes/group.pm:41]: I know what you mean. I did calculate it twice here, but I verified that setting it to a variable didn't work, so I changed it back . This is the problem I encountered when I was working with group objects today . From the verification results, I have changed the right one, but the principle has to wait for me to learn before I can understand.

Comment (nkh) [reply lib/App/Asciio/stripes/group.pm:41]: here is an example:

my @l = map { my $x = $_ * 2 ; $x + $x + $x ; } (1, 2, 3) ;
print "@l\n" ;

as you can see $x is computed from $_ and then re-used multiple times.


also note that the use of map is a bit "not so good), map's result should be re-used, it's not a for loop as we use it in this code.

the code would be MUCH better using max()

something like:

$width = max( map { usc_length $_ } split("\n", $text) ;

File: lib/App/Asciio/Elements.pm [hunk 4]
```diff
@@ -572,50 +572,20 @@ delete $self->{SELECTION_INDEX} unless $self->get_selected_elements(1) ;
...
-	if(exists $element->{GROUP} && defined $element->{GROUP}[-1])
-		{
-		$groups_to_select{$element->{GROUP}[-1]}++ ;
-		}
-	}
+my $search_words = $self->display_edit_dialog("input search words", '', $self);
 
-# select groups
 for my $element (@{$self->{ELEMENTS}}) 
 	{
```
Comment (nkh) [inline lib/App/Asciio/Elements.pm:582]: This is a better implementation:

it's half the length and all the selection details are hidden (and probably better handled)

for my $element (@{$self->{ELEMENTS}}) 
	{
	$self->select_elements(1, $element)
		if $self->transform_elements_to_ascii_buffer($element) =~ m/$search_words/i ;
	}

File: lib/App/Asciio/Elements.pm [hunk 5]
```diff
@@ -580,8 +580,7 @@ my $search_words = $self->display_edit_dialog("input search words", '', $self);
 
 for my $element (@{$self->{ELEMENTS}}) 
 	{
-	my $text = $self->transform_elements_to_ascii_buffer($element);
-	if($text =~ m/$search_words/i)
+	if($self->transform_elements_to_ascii_buffer($element) =~ m/$search_words/i)
 		{
 		$element->{SELECTED} = ++$self->{SELECTION_INDEX} ;
```
Comment (nkh) [inline lib/App/Asciio/Elements.pm:585]: please read the comment I wrote before, taking only half of what I wrote is not good enough

Comment (qindapao) [reply lib/App/Asciio/Elements.pm:585]: Sorry, I'll add another submission to the other rebase branch tomorrow. I just saw your comments. I'm ready to rest. Good Night

**Pull request closed without merging**